Dockerfile

```diff
@@ -12,7 +12,16 @@ WORKDIR /app
 # build typed-ast in separate stage because it requires gcc and libc-dev
 FROM base AS python-deps
 RUN apt-get update -qqy && apt-get install -qqy gcc libc-dev
-COPY requirements.txt ./
+
+# install poetry
+RUN pip install --no-cache-dir --quiet poetry
+
+# copy poetry files over
+COPY ./pyproject.toml ./poetry.lock ./
+
+# Just need the requirements.txt from Poetry
+RUN poetry export --no-interaction --output requirements.txt --without-hashes
+
 # use --no-deps to work around https://github.com/pypa/pip/issues/9644
 RUN pip install --no-deps -r requirements.txt
```

Comment:
In my experience, `pip install` with requirements output from poetry is very shaky. Here's how you can do this without pip.

```Dockerfile
ARG PYTHON_VERSION=3.10

FROM python:${PYTHON_VERSION}-slim-buster as base

ENV PYTHONUNBUFFERED=1 \
    PYTHONFAULTHANDLER=1 \
    PYSETUP_PATH="/opt/pysetup" \
    VENV_PATH="/opt/pysetup/.venv" \
    POETRY_HOME="/opt/poetry" \
    POETRY_VIRTUALENVS_IN_PROJECT=true \
    POETRY_NO_INTERACTION=1

ENV PATH="$POETRY_HOME/bin:$VENV_PATH/bin:$PATH"

FROM base as builder

ARG POETRY_VERSION=1.2.1

WORKDIR $PYSETUP_PATH
COPY . .

RUN apt-get update &&\
    apt-get install -y --no-install-recommends curl
RUN curl -sSL https://install.python-poetry.org | python - --version $POETRY_VERSION &&\
    poetry install -vvv --no-dev

FROM base as final
ARG USER=example
ARG HOME_DIR=/home/$USER
ARG PROJECT_DIR=$HOME_DIR/src

# create user with home directory and bash as the default shell
RUN groupadd --gid 5000 $USER
RUN useradd --home-dir $HOME_DIR --create-home --uid 5000 \
      --gid 5000 --shell /bin/bash --skel /dev/null $USER

WORKDIR $PROJECT_DIR

# copy python dependencies from builder layer
COPY --from=builder --chown=5000:5000 $VENV_PATH $VENV_PATH

RUN chown 5000:5000 $PROJECT_DIR

# use container as non-root
USER 5000
```